### PR TITLE
Sparse matrices interface parallel

### PR DESCRIPTION
**Pull request by RiccardoRossi**
**📝 Description**
Adds basic python interface for distributed sparse matrices (and related objects)

Please mark the PR with appropriate tags: 
- Api Breaker, Mpi, etc...

**🆕 Changelog**
Adds basic python interface for distributed sparse matrices (and related objects). 
The goal is to enable basic manipulation from the python layer, so to eventually allow simple FEM assembly from the python layer as well as experimenting on preconditioners together with the scipy sparse interface

### CODE REVIEW

File: kratos/mpi/tests/test_distributed_sparse_matrices.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        nonlocal_data=importer.ImportData(b) #here is where communications happen
+        nonlocal_output = KratosMultiphysics.Vector(nonlocal_mat.size1())
+        nonlocal_output.fill(0.0)
+        nonlocal_mat.SpMV(nonlocal_data, nonlocal_output)
+
+        output = local_output + nonlocal_output
+
+        for i in range(y.GetLocalData().Size()):
+            self.assertEqual(y.GetLocalData()[i],  output[i], 1e-14 )
+
```
Comment (loumalouomega) [inline kratos/mpi/tests/test_distributed_sparse_matrices.py:159]: Many whitelines....

File: kratos/mpi/tests/test_distributed_sparse_matrices.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        return local_connectivities
+
+    def test_matrix_construction(self):
+        kratos_comm  = KratosMultiphysics.DataCommunicator.GetDefault()
+
+        world_size = kratos_comm.Size();
+        my_rank = kratos_comm.Rank();
+
+        dofs_bounds = self.ComputeBounds(40, world_size, my_rank)
+        #auto reference_A_map = DistTestingInternals::GetReferenceMatrixAsMap(dofs_bounds);
```
Comment (loumalouomega) [inline kratos/mpi/tests/test_distributed_sparse_matrices.py]: This is C++ code
```suggestion
```

File: kratos/mpi/tests/test_distributed_sparse_matrices.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+        el_bounds = self.ComputeBounds(31, world_size, my_rank)
+        #const auto connectivities = DistTestingInternals::ElementConnectivities(el_bounds);
```
Comment (loumalouomega) [inline kratos/mpi/tests/test_distributed_sparse_matrices.py]:
```suggestion
```
C++ code

File: kratos/mpi/tests/test_distributed_sparse_matrices.py (hunk 1 above)
Comment (loumalouomega) [reply kratos/mpi/tests/test_distributed_sparse_matrices.py:159]: Remove extra white lines

**Review by loumalouomega — COMMENTED** [review]
Some comments

**Comment by philbucher** [comment]
@RiccardoRossi the main comment I have is if it can be refactored to reuse #9542 or at least parts of it
can you comment please?
I see that they are not 100% matching but maybe it is possible to refactor some stuff

**Comment by RiccardoRossi** [comment]
Hi @philbucher i don't think you can refactor it more... the two implementations have a very similar interface (at least as similar as i could make it) but the one of distributed is quite more extensive and relies on a number of auxiliary objects that are simply not there ins the serial case (for example DistributedNumbering). The number of copied functions is limited to a very few lambdas that are necessary to solve the overloads (...if we had at least c++14 we could get rid of those by using pybind's overlading helpers)

The point is that i want the python interface to be sufficiently flexible that someone can experiment with preconditioning on the top of it, and this requires exposing quite a few helper functions.
for example if you see the test i made i "manually" did the parallel SpMV on the basis of such low level interface.

File: kratos/mpi/python/add_mpi_sparse_matrix_interface_to_python.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main author:     Jordi Cotela
```
Comment (philbucher) [inline kratos/mpi/python/add_mpi_sparse_matrix_interface_to_python.h]:
```suggestion
//  Main author:     Riccardo Rossi
```

File: kratos/mpi/tests/test_distributed_sparse_matrices.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+import KratosMultiphysics.KratosUnittest as KratosUnittest
+import KratosMultiphysics
+import KratosMultiphysics.mpi
+from KratosMultiphysics.testing.utilities import ReadModelPart
+import numpy as np
+
+def GetFilePath(fileName):
+    return os.path.dirname(os.path.realpath(__file__)) + "/" + fileName
+
+class TestDistributedSparseMatrices(KratosUnittest.TestCase):
```
Comment (philbucher) [inline kratos/mpi/tests/test_distributed_sparse_matrices.py:11]: does this test run with any number of cores?
Or do we have to skip it e.g. if too many cores are used?

File: kratos/mpi/python/add_distributed_sparse_matrices_to_python.cpp [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+{
+    namespace py = pybind11;
+    typedef std::size_t IndexType;
+    typedef std::size_t MPIIndexType;
+
+    py::class_<DistributedNumbering<IndexType>,DistributedNumbering<IndexType>::Pointer>(m,"DistributedNumbering")
+        .def(py::init<const DataCommunicator&, IndexType>())
+        .def(py::init<const DataCommunicator&, IndexType, MPIIndexType>())
+        .def("GetComm", &DistributedNumbering<IndexType>::GetComm)
+        .def("pGetComm", &DistributedNumbering<IndexType>::pGetComm)
```
Comment (philbucher) [inline kratos/mpi/python/add_distributed_sparse_matrices_to_python.cpp]: pGet... in the python interface?

File: kratos/mpi/python/add_distributed_sparse_matrices_to_python.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    py::class_<DistributedVectorImporter<double,IndexType>,DistributedVectorImporter<double,IndexType>::Pointer>(m,"DistributedVectorImporter")
+        .def(py::init<const DataCommunicator&, const std::vector<IndexType>&, const DistributedNumbering<IndexType>&>())
+        .def("GetComm", &DistributedVectorImporter<double,IndexType>::GetComm)
+        .def("ImportData", &DistributedVectorImporter<double,IndexType>::ImportData)
+        ;
+
+    py::class_<DistributedSparseGraph<IndexType>, DistributedSparseGraph<IndexType>::Pointer>(m,"DistributedSparseGraph")
+        .def(py::init<const IndexType, DataCommunicator&>())
+        .def("GetComm", &DistributedSparseGraph<IndexType>::GetComm)
+        .def("pGetComm", &DistributedSparseGraph<IndexType>::pGetComm)
```
Comment (philbucher) [inline kratos/mpi/python/add_distributed_sparse_matrices_to_python.cpp]: same

File: kratos/mpi/python/add_distributed_sparse_matrices_to_python.cpp [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        })
+        .def("__str__", PrintObject<DistributedSparseGraph<IndexType>>)
+        ;
+
+    py::class_<DistributedSystemVector<double,IndexType>,  DistributedSystemVector<double,IndexType>::Pointer>(m,"DistributedSystemVector")
+        .def(py::init<const DistributedSparseGraph<IndexType>&>())
+        .def(py::init<DistributedSystemVector<double,IndexType>&>())
+        .def(py::init<const DistributedNumbering<IndexType>&>())
+        .def("GetComm", &DistributedSystemVector<double,IndexType>::GetComm)
+        .def("pGetComm", &DistributedSystemVector<double,IndexType>::pGetComm)
```
Comment (philbucher) [inline kratos/mpi/python/add_distributed_sparse_matrices_to_python.cpp]: same

File: kratos/mpi/python/add_distributed_sparse_matrices_to_python.cpp [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        .def("__setitem__", [](DistributedSystemVector<double,IndexType>& self, const unsigned int i, const double value){self[i] = value;} )
+        .def("__getitem__", [](const DistributedSystemVector<double,IndexType>& self, const unsigned int i){return self[i];} )
+        .def("__str__", PrintObject<DistributedSystemVector<double,IndexType>>)
+        ;
+
+
+    py::class_<DistributedCsrMatrix<double,IndexType>, DistributedCsrMatrix<double,IndexType>::Pointer>(m,"DistributedCsrMatrix")
+        .def(py::init<const DistributedSparseGraph<IndexType>&>())
+        .def("GetComm", &DistributedCsrMatrix<double,IndexType>::GetComm)
+        .def("pGetComm", &DistributedCsrMatrix<double,IndexType>::pGetComm)
```
Comment (philbucher) [inline kratos/mpi/python/add_distributed_sparse_matrices_to_python.cpp]: same

**Review by philbucher — COMMENTED** [review]
I was wondering why you added the `pGetComm` in the python interface (it is also there in the serial version)

File: kratos/mpi/python/add_distributed_sparse_matrices_to_python.cpp (hunk 9 above)
Comment (RiccardoRossi) [reply kratos/mpi/python/add_distributed_sparse_matrices_to_python.cpp]: ok, removed (to all of the previous)

File: kratos/mpi/tests/test_distributed_sparse_matrices.py (hunk 5 above)
Comment (RiccardoRossi) [reply kratos/mpi/tests/test_distributed_sparse_matrices.py:11]: it should run with any number of cores, yes

**Review by philbucher — APPROVED** [review]
cool!

**Pull request merged**